src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -408,14 +408,13 @@ protected BoundExpression MakeRelationalTest(SyntaxNode syntax, BoundExpression
                 Debug.Assert(loweredInput.Type is { });
 
                 // We share input variables if there is no when clause (because a when clause might mutate them).
-                bool anyWhenClause =
-                    decisionDag.TopologicallySortedNodes
-                    .Any(node => node is BoundWhenDecisionDagNode { WhenExpression: { ConstantValue: null } });
+                bool anyWhenClause = decisionDag.TopologicallySortedNodes.Any(node => isWhenClause(node));
 
                 var inputDagTemp = BoundDagTemp.ForOriginalInput(loweredInput);
                 if ((loweredInput.Kind == BoundKind.Local || loweredInput.Kind == BoundKind.Parameter)
                     && loweredInput.GetRefKind() == RefKind.None &&
-                    !anyWhenClause)
+                    !anyWhenClause &&
+                    !decisionDag.TopologicallySortedNodes.Any(node => isNotClause(node)))
```

Comment:
>!decisionDag.TopologicallySortedNodes.Any(node => isNotClause(node)) [](start = 20, length = 68)

It is quite possible I am missing something here, but, given the comment inside the ```if```, it feels like addition of this condition goes against the intent of the modified ```if``` condition. Specifically, the presence of the "is not" clause doesn't indicate a possibility for the input variable to change, therefore, the optimization should still be valid. It feels like there problem is elsewhere and we simply failing to "and we just reuse the local variable when we need the input expression" (i.e. failing to take optimization into account) when we generate code for the pattern failure branch.

Reply:
> failing to take optimization into account) when we generate code for the pattern failure branch

I believe we do not generate code at the end of failure branches, only success branches.

---
In reply to: [524447959](https://github.com/dotnet/roslyn/pull/49369#discussion_r524447959) [](ancestors = 524447959)

Reply:
> I believe we do not generate code at the end of failure branches, only success branches.

Obviously somewhere we refer to the local and not taking optimization into account

---
In reply to: [525428080](https://github.com/dotnet/roslyn/pull/49369#discussion_r525428080) [](ancestors = 525428080,524447959)